pex/cli/commands/docs.py

```diff
@@ -0,0 +1,220 @@
+                for line in fp:
+                    if line.endswith(os.linesep):
+                        match = re.search(r"Serving HTTP on 0.0.0.0 port (?P<port>\d+)", line)
+                        if match:
+                            port = match.group("port")
+                            return "http://localhost:{port}".format(port=port)
+        return None
+
+    @classmethod
+    def record(
+        cls,
+        server_log,  # type: str
+        pid,  # type: int
+        timeout=5.0,  # type: float
+    ):
+        # type: (...) -> Optional[Pidfile]
+        url = cls._read_url(server_log, timeout)
+        if not url:
+            return None
+
+        with safe_open(cls._PIDFILE, "w") as fp:
+            json.dump(dict(url=url, pid=pid), fp, indent=2, sort_keys=True)
+        return cls(url=url, pid=pid)
+
+    url = attr.ib()  # type: str
+    pid = attr.ib()  # type: int
+
+    def alive(self):
+        # type: () -> bool
+        # TODO(John Sirois): Handle pid rollover
+        try:
+            os.kill(self.pid, 0)
+            return True
+        except OSError as e:
+            if e.errno == errno.ESRCH:  # No such process.
+                return False
+            raise
+
+    def kill(self):
+        # type: () -> None
+        os.kill(self.pid, signal.SIGTERM)
+
+
+@attr.s(frozen=True)
+class LaunchResult(object):
+    url = attr.ib()  # type: str
+    already_running = attr.ib()  # type: bool
+
+
+def launch_docs_server(
+    document_root,  # type: str
+    port,  # type: int
+    timeout=5.0,  # type: float
+):
+    # type: (...) -> Union[str, LaunchResult]
+
+    pidfile = Pidfile.load()
+    if pidfile and pidfile.alive():
+        return LaunchResult(url=pidfile.url, already_running=True)
+
+    # Not proper daemonization, but good enough.
+    log = os.path.join(SERVER_DIR, "log.txt")
+    http_server_module = "http.server" if sys.version_info[0] == 3 else "SimpleHttpServer"
+    env = os.environ.copy()
+    # N.B.: We set up line buffering for the process pipes as well as the underlying Python running
+    # the http server to ensure we can observe the `Serving HTTP on ...` line we need to grab the
+    # ephemeral port chosen.
+    env.update(PYTHONUNBUFFERED="1")
+    with safe_open(log, "w") as fp:
+        process = subprocess.Popen(
+            args=[sys.executable, "-m", http_server_module, str(port)],
+            env=env,
+            cwd=document_root,
+            preexec_fn=os.setsid,
+            bufsize=1,
+            stdout=fp.fileno(),
+            stderr=subprocess.STDOUT,
+        )
+
+    pidfile = Pidfile.record(server_log=log, pid=process.pid, timeout=timeout)
+    if not pidfile:
+        try:
+            os.kill(process.pid, signal.SIGKILL)
+        except OSError as e:
+            if e.errno != errno.ESRCH:  # No such process.
+                raise
+        return log
+
+    return LaunchResult(url=pidfile.url, already_running=False)
+
+
+def shutdown_docs_server():
```

Comment:
Is this function unused?

Reply:
Yeah. This will be used in a follow-up I think; see my reply to your idea of blocking and expect user to know how to ~background a job.